### PR TITLE
Make themes DPI aware

### PR DESCRIPTION
**Pull request by jozefizso**
I've updated theme loader to scale UI elements at load time to the DPI level of primary monitor.
I've ported the splashcreen code from https://github.com/wixtoolset/wix4/pull/46

The burn.exe is marked as DPI aware. Loading the DPI of primary monitor should be sufficient for general use. I could not make the rescaling based on current monitor. I tested the built-in themes and they worked correctly.

I also updated thmviewer.exe to support high DPI monitors.

There may be some quirks in the implementation. This fixes the https://github.com/wixtoolset/issues/issues/3920

### CODE REVIEW

**Comment by rseanhall** [comment]
Thanks for the pull request.  Although this bug was triaged for v3.x, the scope of this means that we need to get it into v4 first (this is reminding us that the Burn engine has GDI+ in v4 and we might not have fully tested that for the DLL hijacking issue, plus marking the exe as high-DPI aware might cause display issues in existing BA's).  Please submit a new pull request for this in the wix4 repo.

A quick glance over the code shows a couple of style nits.  According to the [style guide](http://wixtoolset.org/development/code-style/), all local variables should be declared and initialized at the top of the function and there should be only one return statement from a function.

**Comment by barnson** [comment]
I'd rather not take GDI+ into the engine in v3. Other than splash screen, is there any hidpi functionality we lose without it? The ThmUtil change is great; I'll review that part soon(ish).

**Comment by jozefizso** [comment]
The high DPI works even without that change to splash screen so this should not require GDI+.

**Comment by glopesdev** [comment]
Any updates on this pull request? It would be great to have high DPI support on wix v3, given the fast increasing number of devices that make use of it. Given that apparently no GDI+ support is needed, is there anything else holding this back?

File: src/libs/dutil/gdiputil.cpp [hunk 1]
```diff
@@ -210,3 +210,55 @@ HRESULT DAPI GdipHresultFromStatus(
 
     return E_UNEXPECTED;
 }
+
+/********************************************************************
+GetDpiForMonitor - returns the X and Y DPI values for specified monitor.
+
+********************************************************************/
+BOOL DAPI GetDpiForMonitor(
```
Comment (barnson) [inline src/libs/dutil/gdiputil.cpp]: This doesn't really belong in GdipUtil. How about a new DpiUtil? Make this `DpiGetForWindow`?

File: src/libs/dutil/gdiputil.cpp [hunk 2]
```diff
@@ -210,3 +210,55 @@ HRESULT DAPI GdipHresultFromStatus(
...
+    *nDpiY = 96;
+
+    return FALSE;
+}
+
+/********************************************************************
+GetScaleFactorForDpi - returns the scale factor for given DPI resolution.
+
+********************************************************************/
+FLOAT DAPI GetScaleFactorForDpi(
```
Comment (barnson) [inline src/libs/dutil/gdiputil.cpp]: `DpiGetScaleFactor`?

File: src/libs/dutil/inc/thmutil.h [hunk 3]
```diff
@@ -550,6 +554,11 @@ DAPI_(void) ThemeSetFocus(
     __in DWORD dwControl
     );
 
+DAPI_(int) ScaleByFactor(
```
Comment (barnson) [inline src/libs/dutil/inc/thmutil.h]: Does this need to be public? If so, needs a `Thm` prefix.

**Comment by barnson** [comment]
Oh, and can you add a `.md` file in the `history` directory instead of editing `history.md` directly? That'll avoid the merge conflict.

**Comment by rakeshsinghranchi** [comment]
@barnson ,What version of wixtoolset should one use to consume this fix ? Thanks.

**Comment by barnson** [comment]
It hasn't been merged.

**Comment by rakeshsinghranchi** [comment]
Is there a timeline planned to merge this fix?

**Comment by livarcocc** [comment]
pinging on this thread. When will we have a wixtoolset version with this fix in it?

**Comment by barnson** [comment]
The engine changes are already present in WiX v4. The ThmUtil changes can be merged in WiX v4 when the open issues are addressed.

**Comment by rakeshsinghranchi** [comment]
cc @merriemcgaw

**Comment by johnbeisner** [comment]
@robmen 
@barnson 
A comment from https://github.com/wixtoolset/issues/issues/3920 related to Hi-DPI support of themes:
"Bug is open in v3.x which means it's waiting for someone to step up and declare interest to fix it in a particular WiX v3 release."

And I see that someone has indeed stepped up: 
@jozefizso 
PR: https://github.com/wixtoolset/wix3/pull/376

When can we see Hi-DPI support in WiX 3 or WiX 4? Can we get a general ETA so we can plan accordingly.

**Comment by barnson** [comment]
As I mentioned before, there are still open issues so it hasn't been merged. This PR is tagged `moveto4` to look at doing that work before WiX v4.0 ships.

**Comment by livarcocc** [comment]
hey @barnson thanks for getting back to us. Where can I find information regarding when WiX v4.0 is shipping? Also, if there is something we can do to help address some of those issues, would love to help.

**Comment by barnson** [comment]
@livarcocc, WiX v4 is still in active development, so we don't have useful dates yet beyond "this year." If you wanted to port the change in this PR to https://github.com/wixtoolset/dutil and address the issues I raised above, they'd be available sooner than "when I have the spare time."

**Pull request closed without merging**